RedHatInsights/rhsm-subscriptions
# ENT-3871/ENT-3978: Update metrics task to use tag profile

This PR combines changes for:
[ENT-3871](https://issues.redhat.com/browse/ENT-3871)
[ENT-3978](https://issues.redhat.com/browse/ENT-3978)

### Added a PromQL query builder

The QueryBuilder uses Spring Expression Language (SPEL) to apply
variables to a configured query template.

A query template is expected to be a well formed PromQL query.
Query templates are to be defined as a map (templateKey: promql)
in a configurationfile and can be be accessed via
PrometheusMetricsProperties:

   rhsm-subscriptions.metering.prometheus.metric.queryTemplates

A template can include template variables via a SPEL expressions.
These variables are substituted when the query is built and template
values are defined in the QueryDescriptor passed to the builder
at build time. Every property defined in the descriptor object will
accessible as template variables when the query is built. Variable
names are specified in dot notation and can be nested to access any
value defiend in the descriptor.

A QueryDescriptor contains two key fields that can be accessed in
a query template.
* **tag**: Comes from a configured TagMetric
* **runtime**: A generic collection of runtime args passed at runtime.

While the example below is not PromQL, an example of accessing
descriptor values would be as follows:

```
Metric #{tag.metricId} looked up for Account #{runtime[account]}
```

NOTE:
Template variables can contain other SPEL expressions, but it is important
note that we support 3 levels of nesting out of the box. This is not a
supported feature of SPEL. The number of levels we support was made
configurable, but I can't see us needing to change it.

###  Update metrics task to dynamically run query

* When the MeteringJob runs, pull the supported tags
  and UOMs from the tag profile definition.
* Defined a default query template for metrics.
* Add query parameters to TagMetric definitions that automatically
  get applied to the prometeus query based on the tag/uom when the task runs.
* Apply appropriate tag properties to events when they are generated
  from prometheus metrics.

### Testing
1. Configure access to telemeter.
```
podman run --rm -d -ti -p 8082:8080 quay.io/observatorium/token-refresher:master-2021-02-05-5da9663 --oidc.client-secret=$(cat ~/.token-refresher-stage) --oidc.client-id=observatorium-subwatch-staging --oidc.issuer-url=https://sso.redhat.com/auth/realms/redhat-external --url=https://observatorium.api.stage.openshift.com
```

2. Run the app
```
USER_USE_STUB=true PROM_URL="http://localhost:8082/api/metrics/v1/telemeter/api/v1" DEV_MODE=true ./gradlew clean bootRun
```

3. Kick off metrics gathering for OpenShift-metrics and observe in the logs that a task is created for each account for metering Cores. Verify that events are created.
```
curl 'http://localhost:8080/actuator/hawtio/jolokia/' -H 'Content-Type: application/json' -d '{"type":"exec","mbean":"org.candlepin.subscriptions.metering.jmx:name=meteringJmxBean,type=MeteringJmxBean","operation":"performCustomMetering(java.lang.String, java.lang.String, java.lang.Integer)","arguments":["OpenShift-metrics", null, 1440]}'
```

4. In an ideal test environment, data would exist in prometheus that would allow us to properly test each product, but sadly in stage prometheus there is currently only data for OpenShift-metrics. We can test that Events are getting created for OpenShift-dedicated-metrics though, if we modify the tag_profile.yaml. Modify the yaml file and kick off metrics gathering for OpenShift-dedicated-metrics and observe in the logs that a task is created for each account for metering BOTH Cores AND Instance-hours. Verify that events are created for cores and instance-hours.
```yaml
# Modify the prometheusMetadataMetric query param in tag_profile.yaml so that the product filter is osd

# OSD metrics
- tag: OpenShift-dedicated-metrics
    metricId: redhat.com:openshift_dedicated:4cpu_hour
    uom: CORES
    queryParams:
      prometheusMetric: cluster:usage:workload:capacity_physical_cpu_cores:max:5m
      prometheusMetadataMetric: subscription_labels{product="ocp", ebs_account="#{runtime[account]}", billing_model="marketplace", support=~"Premium|Standard|Self-Support|None"}
  - tag: OpenShift-dedicated-metrics
    metricId: redhat.com:openshift_dedicated:cluster_hour\
    uom: INSTANCE_HOURS
    queryParams:
      prometheusMetric: group(cluster:usage:workload:capacity_physical_cpu_cores:max:5m) by (_id)
      prometheusMetadataMetric: subscription_labels{product="ocp", ebs_account="#{runtime[account]}", billing_model="marketplace", support=~"Premium|Standard|Self-Support|None"}
```
```
curl 'http://localhost:8080/actuator/hawtio/jolokia/' -H 'Content-Type: application/json' -d '{"type":"exec","mbean":"org.candlepin.subscriptions.metering.jmx:name=meteringJmxBean,type=MeteringJmxBean","operation":"performCustomMetering(java.lang.String, java.lang.String, java.lang.Integer)","arguments":["OpenShift-dedicated-metrics", null, 1440]}'
```

Tip:
Query the event data as follows to clean up the output.
```sql
select instance_id, account_number, event_type, data->>'sla', data->>'role', data->>'usage', data->>'measurements' from events;
```
